torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1196,6 +1213,329 @@ def call_function(
         return super().call_function(tx, args, kwargs)
 
 
+class CannotConvertRangeToHigherOrder(TorchDynamoException):
+    pass
+
+
+class RangeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    The general idea is that for a simple for loop like
+
+    ```
+    for i in range(10):
+        x = x + 1
+    ```
+
+    We can convert the loop body to a function and the loop itself
+    to a series of calls:
+
+    ```
+    def loop_body(local_0, local_1, local_...):
+        local_1 = local_1 + 1
+        return (local_0, local_1, local_...)
+
+    (i, x) = loop_body(0, x)
+    (i, x) = loop_body(1, x)
+    (i, x) = loop_body(2, x)
+    # ... 7 more times
+    ```
+
+    There are a few benefits to doing this, but mainly we bypass a ton
+    of stuff associated with tracing for loops. That itself provides some speedup.
+
+    Once we implement function compilation caching, or a higher order operator,
+    like jax.lax.fori_loop, for loops will compile in constant time, and perhaps even
+    better than what we currently have.
+    """
+
+    # Any of these nested means the control flow is too complex to be
+    # represented by a higher order op.
+    FORBIDDEN_OPCODES = {
```

Comment:
How difficult would it be to reverse this? That is, have a list of allowed opcodes instead of a list of forbidden opcodes.

The reason I ask because this doesn't seem very robust; if python adds a new opcode for something weird then it automatically gets allowed and we don't have good testing for it.

Reply:
I just tried and the list of opcodes is too long. We would need to list all opcodes from 3.8-3.10, and it changed once in 3.11, so that too, plus slight changes again in 3.12.

Reply:
We can also abort early if we find an opcode in the loop body that can't be traced with symbolic_convert.

Reply:
Let's add this check as @williamwen42 suggested